roofit/roofitcore/src/RooLinkedList.cxx

```diff
@@ -615,13 +616,14 @@ TObject* RooLinkedList::find(const char* name) const
       if (nptr && nptr->TestBit(RooNameReg::kRenamedArg)) {
         RooLinkedListElem* ptr = _first ;
         while(ptr) {
-          if ((((RooAbsArg*)ptr->_arg)->namePtr() == nptr)) {
+          if (static_cast<RooAbsArg*>(ptr->_arg)->namePtr() == nptr ||
+              static_cast<RooAbsData*>(ptr->_arg)->namePtr() == nptr) {
```

Comment:
```suggestion
          if (static_cast<RooAbsArg*>(ptr->_arg)->namePtr() == nptr ||
              static_cast<RooAbsData*>(ptr->_arg)->namePtr() == nptr) {
```
I think that's undefined behavior here! What if the name pointer comparison is used because `!dynamic_cast<RooAbsArg*>(arg) && !dynamic_cast<RooAbsData*>(arg)` was true, and you have a `RooAbsData*`? In that case, `static_cast<RooAbsArg*>(ptr->_arg)->namePtr()` is wrong.

You could do again `dynamic_cast`, but this is expensive and redundant because you already do it before. But there are other solution. Maybe further up in this function, when you do the dynamic_cast, you can assign a local pointer variable with the correct name pointer?

Reply:
Probably I don't understand casts well enough.

`_useNptr` is `false` when any of the elements added to the list is not dynamic_cast-able to either `RooAbsArg` or `RooAbsData`.

Then in `find` if `useNptr` is `true` then we know `_arg` should be dynamic_cast-able to either of the two classes. I am not able to appreciate the cases where an argument would be dynamic_cast-able but not static_cast-able.

Maybe the right thing is to use static_cast in the first instance (i.e when adding objects) ?

In any case, please note the code I am extending and that has been used for years uses dynamic_cast in `Add` and C-style cast (that I am replacing with static_cast) in `find`

Reply:
> Then in `find` if `useNptr` is `true` then we know `_arg` should be dynamic_cast-able to either of the two classes. I am not able to appreciate the cases where an argument would be dynamic_cast-able but not static_cast-able.

From the `dynamic_cast`s, we know that `_arg` can only be downcasted to *only one* of `RooAbsArg` or `RooAbsData`. So you can't just `static_cast` to both.

> In any case, please note the code I am extending and that has been used for years uses dynamic_cast in `Add` and C-style cast (that I am replacing with static_cast) in `find`

That's true, but before it was clear that `_arg` would be a RooAbsArg if the `dynamic_cast` succeeded, so the C-style cast was valid. Now you don't know anymore what it is, changing the picture completely.

Reply:
I think you understand the casts just right, but what changed that the collection can now contain two different types of objects:
Assume it's an `AbsData`-derived object. Now, you run a static_cast to `RooAbsArg*` on it, and immediately call the function `RooAbsArg::namePtr()` on it. This will crash (or not), because the `_namePtr` member is not at the same offset in `AbsArg` as it is in `AbsData`.
So, to expand on what Jonas meant with dynamic_cast:
```c++
if ( (dynamic_cast<RooAbsArg*>(ptr->_arg) && static_cast<RooAbsArg*>(ptr->_arg)->namePtr() == nptr) || 
     (dynamic_cast<RooAbsData*>(ptr->_arg) && static_cast<RooAbsData*>(ptr->_arg)->namePtr() == nptr) )
```
should do it, as it first checks the type and only calls the function if it's safe.

Reply:
Many thanks to both of you for the explanations, now I get it !

I think the solution proposed by @hageboeck is the right way to go, at least for now. The case in this `if` should be triggered only very rarely, as it applies only to objects that have been renamed.

Reply:
Thanks a lot, also to Stephan for clarifying! Using the wrong offset for the `_namePtr` member was a ticking time bomb for address boundary error :slightly_smiling_face: